File: strategy-calc.js
// 策略計算引擎:在瀏覽器計算所有技術指標
function sma(arr, period, idx) {
  if (idx === undefined) idx = arr.length - 1;
  if (idx < period - 1) return null;
  let sum = 0;
  for (let i = idx - period + 1; i <= idx; i++) sum += arr[i];
  return sum / period;
}
function ema(arr, period) {
  const k = 2 / (period + 1);
  const out = [];
  let prev;
  arr.forEach((v, i) => {
    if (i === 0) { prev = v; out.push(v); }
    else { prev = v * k + prev * (1 - k); out.push(prev); }
  });
  return out;
}
function mansfieldRS(closes, benchCloses, length = 50, smooth = 5) {
  const rsRatio = closes.map((c, i) => benchCloses[i] ? c / benchCloses[i] : null);
  const mansfield = rsRatio.map((r, i) => {
    const s = sma(rsRatio, length, i);
    return (r && s) ? ((r / s) - 1) * 100 : null;
  });
  const valid = mansfield.map(v => v === null ? 0 : v);
  const smoothed = ema(valid, smooth);
  return { mansfield, smoothed };
}
function tdSequential(closes) {
  let sc = 0, bc = 0;
  const scArr = [], bcArr = [];
  for (let i = 0; i < closes.length; i++) {
    if (i >= 4) {
      sc = closes[i] > closes[i - 4] ? sc + 1 : 0;
      bc = closes[i] < closes[i - 4] ? bc + 1 : 0;
    }
    scArr.push(sc); bcArr.push(bc);
  }
  return { sc, bc, scArr, bcArr };
}
function bollinger(closes, period = 20, mult = 2) {
  const idx = closes.length - 1;
  const mid = sma(closes, period, idx);
  if (mid === null) return null;
  let variance = 0;
  for (let i = idx - period + 1; i <= idx; i++) variance += (closes[i] - mid) ** 2;
  const sd = Math.sqrt(variance / period);
  return { mid, upper: mid + mult * sd, lower: mid - mult * sd };
}
function rsi(closes, period = 14) {
  if (closes.length < period + 1) return null;
  let gains = 0, losses = 0;
  for (let i = closes.length - period; i < closes.length; i++) {
    const diff = closes[i] - closes[i - 1];
    if (diff >= 0) gains += diff; else losses -= diff;
  }
  const avgGain = gains / period, avgLoss = losses / period;
  if (avgLoss === 0) return 100;
  const rs = avgGain / avgLoss;
  return 100 - (100 / (1 + rs));
}
function extremeBands(highs, lows, closes, mLen = 23, lkbk = 60) {
  const idx = closes.length - 1;
  const m23 = sma(closes, mLen, idx);
  if (m23 === null) return null;
  let maxU = -Infinity, maxL = -Infinity;
  for (let i = idx - lkbk; i < idx; i++) {
    if (i < mLen - 1) continue;
    const m = sma(closes, mLen, i);
    if (m === null) continue;
    maxU = Math.max(maxU, highs[i] - m);
    maxL = Math.max(maxL, m - lows[i]);
  }
  const h10 = Math.max(...highs.slice(idx - 9, idx + 1));
  const l10 = Math.min(...lows.slice(idx - 9, idx + 1));
  return { m23, sellP: m23 + maxU, buyP: m23 - maxL, h10, l10 };
}
function computeAll(data) {
  const rows = data.target;
  const closes = rows.map(r => r.close);
  const highs = rows.map(r => r.high);
  const lows = rows.map(r => r.low);
  const vols = rows.map(r => r.volume);
  const bench = data.benchmark_close;
  const cur = rows[rows.length - 1];
  const ma20 = sma(closes, 20);
  const ma60 = sma(closes, 60);
  const volMa5 = sma(vols, 5);
  const rsData = mansfieldRS(closes, bench);
  const td = tdSequential(closes);
  const boll = bollinger(closes);
  const rsiVal = rsi(closes);
  const bands = extremeBands(highs, lows, closes);
const monthHigh = Math.max(...highs.slice(Math.max(0, highs.length - 20)));
  return {
    date: cur.date, price: cur.close,
    open: cur.open, high: cur.high, low: cur.low, volume: cur.volume,
    ma20, ma60, volMa5,
    rs: rsData.smoothed[rsData.smoothed.length - 1],
    rsPrev: rsData.smoothed[rsData.smoothed.length - 2],
    sc: td.sc, bc: td.bc,
    boll, rsi: rsiVal, bands, monthHigh,
  };
}
window.computeAll = computeAll;
window.fetchStrategyData = async function() {
  const res = await fetch('data/strategy_data.json?t=' + Date.now());
  return await res.json();
};

// ===== 歷史序列(畫圖用) =====
function computeSeries(data) {
  const rows = data.target;
  const closes = rows.map(r => r.close);
  const highs = rows.map(r => r.high);
  const lows = rows.map(r => r.low);
  const bench = data.benchmark_close;
  const n = rows.length;
  const dates = rows.map(r => r.date);
  const ma20Arr = [], ma60Arr = [], bollUp = [], bollLow = [], bollMid = [];
  const sellPArr = [], buyPArr = [];
  const rsData = mansfieldRS(closes, bench);
  const rsArr = rsData.smoothed;
  for (let i = 0; i < n; i++) {
    ma20Arr.push(sma(closes, 20, i));
    ma60Arr.push(sma(closes, 60, i));
    const mid = sma(closes, 20, i);
    if (mid !== null) {
      let v = 0;
      for (let j = i - 19; j <= i; j++) v += (closes[j] - mid) ** 2;
      const sd = Math.sqrt(v / 20);
      bollMid.push(mid); bollUp.push(mid + 2*sd); bollLow.push(mid - 2*sd);
    } else { bollMid.push(null); bollUp.push(null); bollLow.push(null); }
    const m23 = sma(closes, 23, i);
    if (m23 !== null && i >= 60) {
      let mu = -Infinity, ml = -Infinity;
      for (let j = i - 60; j < i; j++) {
        const m = sma(closes, 23, j);
        if (m === null) continue;
        mu = Math.max(mu, highs[j] - m);
        ml = Math.max(ml, m - lows[j]);
      }
      sellPArr.push(m23 + mu); buyPArr.push(m23 - ml);
    } else { sellPArr.push(null); buyPArr.push(null); }
  }
  const signals = [];
  for (let i = 1; i < n; i++) {
    if (ma60Arr[i] === null) continue;
    const rsUp = rsArr[i] > 0 && rsArr[i-1] <= 0;
    const rsDown = rsArr[i] < 0 && rsArr[i-1] >= 0;
    if (rsUp && closes[i] > ma60Arr[i]) signals.push({ i, type: 'buy', price: closes[i] });
    else if (rsDown && ma20Arr[i] && closes[i] < ma20Arr[i]) signals.push({ i, type: 'sell', price: closes[i] });
  }
  return { dates, closes, ma20Arr, ma60Arr, bollUp, bollLow, bollMid, sellPArr, buyPArr, rsArr, signals };
}
window.computeSeries = computeSeries;

File: scripts/strategy_calc.py
"""策略計算(Python版,供 LINE 推播用)"""

def sma(arr, period, idx=None):
    if idx is None:
        idx = len(arr) - 1
    if idx < period - 1:
        return None
    return sum(arr[idx - period + 1: idx + 1]) / period

def ema(arr, period):
    k = 2 / (period + 1)
    out = []
    prev = None
    for i, v in enumerate(arr):
        prev = v if i == 0 else v * k + prev * (1 - k)
        out.append(prev)
    return out

def mansfield_rs(closes, bench, length=50, smooth=5):
    rs_ratio = [c / b if b else None for c, b in zip(closes, bench)]
    mansfield = []
    for i, r in enumerate(rs_ratio):
        window = rs_ratio[max(0, i-length+1):i+1]
        s = sma(rs_ratio, length, i) if all(x is not None for x in window) else None
        mansfield.append(((r / s) - 1) * 100 if (r and s) else None)
    valid = [v if v is not None else 0 for v in mansfield]
    return ema(valid, smooth)

def td_seq(closes):
    sc = bc = 0
    for i in range(len(closes)):
        if i >= 4:
            sc = sc + 1 if closes[i] > closes[i-4] else 0
            bc = bc + 1 if closes[i] < closes[i-4] else 0
    return sc, bc

def bollinger(closes, period=20, mult=2):
    idx = len(closes) - 1
    mid = sma(closes, period, idx)
    if mid is None:
        return None
    var = sum((closes[i] - mid) ** 2 for i in range(idx - period + 1, idx + 1)) / period
    return {"mid": mid, "upper": mid + mult * var**0.5, "lower": mid - mult * var**0.5}

def rsi(closes, period=14):
    if len(closes) < period + 1:
        return None
    gains = losses = 0
    for i in range(len(closes) - period, len(closes)):
        diff = closes[i] - closes[i-1]
        if diff >= 0:
            gains += diff
        else:
            losses -= diff
    ag, al = gains / period, losses / period
    if al == 0:
        return 100
    return 100 - (100 / (1 + ag / al))

def extreme_bands(highs, lows, closes, m_len=23, lkbk=60):
    idx = len(closes) - 1
    m23 = sma(closes, m_len, idx)
    if m23 is None:
        return None
    max_u = max_l = float('-inf')
    for i in range(idx - lkbk, idx):
        if i < m_len - 1:
            continue
        m = sma(closes, m_len, i)
        if m is None:
            continue
        max_u = max(max_u, highs[i] - m)
        max_l = max(max_l, m - lows[i])
    return {"m23": m23, "sellP": m23 + max_u, "buyP": m23 - max_l,
            "h10": max(highs[idx-9:idx+1]), "l10": min(lows[idx-9:idx+1])}

def compute_signals(data):
    rows = data["target"]
    closes = [r["close"] for r in rows]
    highs = [r["high"] for r in rows]
    lows = [r["low"] for r in rows]
    vols = [r["volume"] for r in rows]
    bench = data["benchmark_close"]
    cur = rows[-1]
    price = cur["close"]
    ma20, ma60, vol_ma5 = sma(closes, 20), sma(closes, 60), sma(vols, 5)
    rs_arr = mansfield_rs(closes, bench)
    rs, rs_prev = rs_arr[-1], rs_arr[-2]
    sc, bc = td_seq(closes)
    boll = bollinger(closes)
    bands = extreme_bands(highs, lows, closes)
    month_high = max(highs[-20:])

    signals = []
    vol_ratio = vols[-1] / vol_ma5 if vol_ma5 else 0
    rs_up = rs > 0 and rs_prev <= 0
    rs_down = rs < 0 and rs_prev >= 0
    ex_sell = bands and cur["high"] >= bands["sellP"] and cur["high"] >= bands["h10"]
    ex_buy = bands and cur["low"] <= bands["buyP"] and cur["low"] <= bands["l10"]
    is_fomo = sc in (8, 9, 12, 13)
    is_panic = bc in (8, 9, 12, 13)
    if ex_sell or is_fomo:
        signals.append("RS訊號:🔴 頂部過熱,禁止追多")
    elif rs_up and price > ma60 and vol_ratio > 1.2:
        signals.append("RS訊號:🟢 轉強買進")
    elif ex_buy or is_panic:
        signals.append("RS訊號:🟢 極端買點/恐慌")
    elif rs_down and price < ma20:
        signals.append("RS訊號:🔴 轉弱賣出")

    if boll and price >= boll["upper"]:
        signals.append("三條線:🔴 觸布林上軌,出清")
    elif boll and price <= boll["lower"] * 1.01:
        signals.append("三條線:🟢 觸布林下軌,第3批")
    elif ma60 and price <= ma60 * 1.01:
        signals.append("三條線:🟢 觸季線,第2批")
    elif ma20 and price <= ma20 * 1.01:
        signals.append("三條線:🟢 觸月線,第1批")

    g1, g2, g3 = month_high * 0.97, month_high * 0.94, month_high * 0.91
    if price <= g3:
        signals.append("網格:🟢 跌破-9%,戰備金全投入")
    elif price <= g2:
        signals.append("網格:🟢 跌破-6%,重倉1.5~2碼")
    elif price <= g1:
        signals.append("網格:🟢 跌破-3%,輕倉0.5碼")

    return {"price": price, "rs": rs, "signals": signals}
